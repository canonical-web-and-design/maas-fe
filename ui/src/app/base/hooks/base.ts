import { useCallback, useEffect, useRef, useState } from "react";

import { NotificationSeverity } from "@canonical/react-components";
import type { NotificationProps } from "@canonical/react-components";
import { useDispatch, useSelector } from "react-redux";

import configSelectors from "app/store/config/selectors";
import { actions as messageActions } from "app/store/message";

/**
 * Add a message in response to a state change e.g. when something is created.
 * @param addCondition - Whether the message should be added.
 * @param cleanup - A cleanup action to fire.
 * @param message - The message to be displayed.
 * @param onMessageAdded - A function to call once the message has
                                      been displayed.
 * @param messageType - The notification type.
 */
export const useAddMessage = (
  addCondition: boolean,
  cleanup: () => { type: string },
  message: string,
  onMessageAdded?: (() => void) | null,
  messageSeverity: NotificationProps["severity"] = NotificationSeverity.INFORMATION
): void => {
  const dispatch = useDispatch();

  useEffect(() => {
    if (addCondition) {
      dispatch(messageActions.add(message, messageSeverity));
      onMessageAdded && onMessageAdded();
      dispatch(cleanup());
    }
  }, [
    addCondition,
    cleanup,
    dispatch,
    message,
    messageSeverity,
    onMessageAdded,
  ]);
};

/**
 * Set the browser window title.
 * @param title - The title to set.
 */
export const useWindowTitle = (title?: string): void => {
  const maasName = useSelector(configSelectors.maasName);
  const maasNamePart = maasName ? `${maasName} ` : "";
  const titlePart = title ? `${title} | ` : "";
  useEffect(() => {
    document.title = `${titlePart}${maasNamePart}MAAS`;
  }, [maasNamePart, titlePart]);
};

/**
 * Handle checking when a value has cycled from false to true.
 * @param value - The value to check.
 * @param onCycled - The function to call when the value changes from false to true.
 */
export const useCycled = (
  value: boolean,
  onCycled?: () => void
): [boolean, () => void] => {
  const previousValue = useRef(value);
  const [hasCycled, setHasCycled] = useState(false);
  useEffect(() => {
    if (value && !previousValue.current) {
      onCycled && onCycled();
      setHasCycled(true);
    }
    if (previousValue.current !== value) {
      previousValue.current = value;
    }
  }, [value, onCycled]);
  const resetHasCycled = useCallback(() => {
    setHasCycled(false);
    previousValue.current = false;
  }, [setHasCycled]);
  return [hasCycled, resetHasCycled];
};

/**
 * Handle displaying action progress for batched actions.
 * @param processingCount - The number of items currently being processed.
 * @param onComplete - The function to call when all the items have been processed.
 * @param hasErrors - Whether there are any item errors in state.
 * @param onError - The function to call when an error occurs.
 */
export const useProcessing = (
  processingCount: number,
  onComplete: () => void,
  hasErrors = false,
  onError: () => void
): void => {
  const processingStarted = useRef(false);
  if (processingStarted.current === false && processingCount > 0) {
    processingStarted.current = true;
  }

  // If all the items have finished processing and there are no errors, run the
  // onComplete function.
  useCycled(processingCount === 0, () => {
    if (!hasErrors) {
      onComplete();
    }
  });

  // If the items are processing and errors occur, run the onError function.
  useCycled(hasErrors, () => {
    if (processingStarted) {
      onError();
    }
  });
};

/**
 * Scroll an element into view on render.
 */
export const useScrollOnRender = <T extends HTMLElement>(): ((
  targetNode: T
) => void) => {
  const htmlRef = useRef<HTMLElement>(document.querySelector("html"));
  const onRenderRef = useCallback((targetNode) => {
    if (targetNode && htmlRef?.current) {
      const { height: targetHeight, y: targetTop } =
        targetNode.getBoundingClientRect();
      const windowTop = htmlRef.current.scrollTop;
      const windowBottom = windowTop + window.innerHeight;
      const targetBottom = targetTop + targetHeight;
      // Whether the top of the target is below the bottom of the screen.
      const topOffBottom = targetTop > windowBottom;
      // Whether the top of the target is above the top of the screen.
      const topOffTop = targetTop < windowTop;
      // Whether the top of the target is on the screen.
      const topOnScreen = !topOffTop && !topOffBottom;
      // Whether the bottom of the target is below the bottom of the screen.
      const bottomOffBottom = targetBottom > windowBottom;
      // Whether the target top is on the screen but the bottom is below the bottom.
      const targetPartiallyOffBottom = topOnScreen && bottomOffBottom;
      if (topOffBottom || topOffTop || targetPartiallyOffBottom) {
        window.scrollTo({
          top: targetTop,
          left: 0,
          behavior: "smooth",
        });
      }
    }
  }, []);
  return onRenderRef;
};
