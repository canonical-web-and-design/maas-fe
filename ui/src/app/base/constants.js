/**
 * Message types defined by MAAS websocket API
 */
export const MESSAGE_TYPES = {
  REQUEST: 0,
  RESPONSE: 1,
  NOTIFY: 2,
};

/**
 * Common column sizes used in <Col> components
 */
export const COL_SIZES = {
  CARD_TITLE: 2,
  DELETE_ROW_BUTTONS: 4,
  SIDEBAR: 3,
  TOTAL: 12,
};

export default MESSAGE_TYPES;
