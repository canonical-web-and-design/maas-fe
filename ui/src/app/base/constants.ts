/**
 * Common column sizes used in <Col> components
 */
export const COL_SIZES = {
  CARD_TITLE: 2,
  TABLE_CONFIRM_BUTTONS: 4,
  SIDEBAR: 3,
  TOTAL: 12,
};

export const COLOURS = {
  CAUTION: "#F99B11",
  LIGHT: "#F7F7F7",
  LINK_FADED: "#D3E4ED",
  LINK: "#0066CC",
  NEGATIVE: "#C7162B",
  POSITIVE_FADED: "#B7CCB9",
  POSITIVE_MID: "#4DAB4D",
  POSITIVE: "#0E8420",
};
