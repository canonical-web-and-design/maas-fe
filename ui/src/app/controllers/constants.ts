import { NodeActions } from "app/store/types/node";

export const ControllerActionHeaderViews = {
  DELETE_CONTROLLER: ["controllerActionForm", NodeActions.DELETE],
  IMPORT_IMAGES_CONTROLLER: ["controllerActionForm", NodeActions.IMPORT_IMAGES],
  OFF_CONTROLLER: ["controllerActionForm", NodeActions.OFF],
  ON_CONTROLLER: ["controllerActionForm", NodeActions.ON],
  OVERRIDE_FAILED_TESTING_CONTROLLER: [
    "controllerActionForm",
    NodeActions.OVERRIDE_FAILED_TESTING,
  ],
  SET_ZONE_CONTROLLER: ["controllerActionForm", NodeActions.SET_ZONE],
  TEST_CONTROLLER: ["controllerActionForm", NodeActions.TEST],
} as const;

export const ControllerNonActionHeaderViews = {
  ADD_CONTROLLER: ["controllerNonActionForm", "addController"],
} as const;

export const ControllerHeaderViews = {
  ...ControllerActionHeaderViews,
  ...ControllerNonActionHeaderViews,
} as const;
