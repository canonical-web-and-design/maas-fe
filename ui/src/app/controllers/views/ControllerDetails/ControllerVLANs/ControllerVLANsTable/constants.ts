export enum ControllerVLANsColumns {
  FABRIC = "fabric",
  VLAN = "vlan",
  DHCP = "dhcp",
  SUBNET = "subnet",
  PRIMARY_RACK = "primary_rack",
  SECONDARY_RACK = "secondary_rack",
}

export const columnLabels = {
  [ControllerVLANsColumns.FABRIC]: "Fabric",
  [ControllerVLANsColumns.VLAN]: "VLAN",
  [ControllerVLANsColumns.DHCP]: "DHCP",
  [ControllerVLANsColumns.SUBNET]: "Subnets",
  [ControllerVLANsColumns.PRIMARY_RACK]: "Primary rack",
  [ControllerVLANsColumns.SECONDARY_RACK]: "Secondary rack",
} as const;
