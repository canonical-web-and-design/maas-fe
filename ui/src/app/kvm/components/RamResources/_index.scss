@mixin RamResources {
  $donut-size: 6.5rem;
  $donut-width: 14px;

  .ram-resources {
    display: flex;
    flex-direction: column;
    overflow: auto;
    padding: $spv-inner--medium $sph-inner;

    .ram-resources__chart-container {
      display: flex;
      flex-direction: column;
    }

    .ram-resources__chart {
      align-self: center;
      width: $donut-size;
    }

    .ram-resources__table {
      margin-bottom: 0;
      margin-top: -$spv-inner--small;
      width: auto;

      th,
      td {
        &:nth-child(1) {
          padding-left: 0;
          width: 50%;
        }

        &:nth-child(2) {
          width: 6.25rem;
        }

        &:nth-child(3) {
          padding-right: 0;
          width: 50%;
        }
      }
    }
  }

  // Additional styles for when the component changes layout depending on the
  // viewport size. Otherwise, the component keeps the "mobile" styling.
  .ram-resources--dynamic-layout {
    @media only screen and (min-width: $breakpoint-medium) {
      flex-direction: row;

      .ram-resources__chart-container {
        flex: 1;
        margin-right: $sph-inner--large;
      }

      .ram-resources__table-container {
        flex: 3;
      }
    }

    @media only screen and (min-width: $breakpoint-large) {
      .ram-resources__chart-container {
        flex: 0;
        margin-right: $sph-inner;
      }

      .ram-resources__chart {
        align-self: flex-start;
      }
    }
  }
}
