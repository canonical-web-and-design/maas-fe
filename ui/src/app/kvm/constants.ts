export const KVMHeaderViews = {
  ADD_LXD_HOST: ["kvmForm", "addLXDHost"],
  ADD_VIRSH_HOST: ["kvmForm", "addVirshHost"],
  COMPOSE_VM: ["kvmForm", "composeVM"],
  DELETE_KVM: ["kvmForm", "deleteKVM"],
  REFRESH_KVM: ["kvmForm", "refreshKVM"],
} as const;
