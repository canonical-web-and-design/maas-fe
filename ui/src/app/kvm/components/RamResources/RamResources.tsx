import classNames from "classnames";

import DoughnutChart from "app/base/components/DoughnutChart";
import { COLOURS } from "app/base/constants";
import { memoryWithUnit } from "app/kvm/utils";

export type Props = {
  dynamicLayout?: boolean;
  general: {
    allocated: number; // B
    free: number; // B
  };
  hugepages: {
    allocated: number; // B
    free: number; // B
    pageSize?: number;
  };
};

const RamResources = ({
  dynamicLayout = false,
  general,
  hugepages,
}: Props): JSX.Element => {
  const totalGeneral = general.allocated + general.free;
  const totalHugepages = hugepages.allocated + hugepages.free;
  const totalMemory = totalGeneral + totalHugepages;
  const overCommitted = general.free < 0 || hugepages.free < 0;

  return (
    <div
      className={classNames("ram-resources", {
        "ram-resources--dynamic-layout": dynamicLayout,
      })}
    >
      <div className="ram-resources__chart-container">
        <h4 className="p-heading--small">RAM</h4>
        <DoughnutChart
          className="ram-resources__chart"
          label={memoryWithUnit(totalMemory)}
          segmentHoverWidth={18}
          segmentWidth={15}
          segments={
            overCommitted
              ? [
                  {
                    color: COLOURS.CAUTION,
                    value: 1,
                  },
                ]
              : [
                  {
                    color: COLOURS.LINK,
                    tooltip: `General allocated ${memoryWithUnit(
                      general.allocated
                    )}`,
                    value: general.allocated,
                  },
                  ...(totalHugepages > 0
                    ? [
                        {
                          color: COLOURS.POSITIVE,
                          tooltip: `Hugepage allocated ${memoryWithUnit(
                            hugepages.allocated
                          )}`,
                          value: hugepages.allocated,
                        },
                        {
                          color: COLOURS.POSITIVE_MID,
                          tooltip: `Hugepage free ${memoryWithUnit(
                            hugepages.free
                          )}`,
                          value: hugepages.free,
                        },
                      ]
                    : []),
                  {
                    color: COLOURS.LINK_FADED,
                    tooltip: `General free ${memoryWithUnit(general.free)}`,
                    value: general.free,
                  },
                ]
          }
          size={96}
        />
      </div>
      <div className="ram-resources__table-container">
        <table className="ram-resources__table">
          <thead>
            <tr>
              <th></th>
              <th className="u-align--right u-text--light">
                <span className="u-nudge-left">Allocated</span>
              </th>
              <th className="u-align--right u-text--light">
                <span className="u-nudge-left">Free</span>
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>General</td>
              <td className="u-align--right">
                {memoryWithUnit(general.allocated)}
                <span className="u-nudge-right--small">
                  <i className="p-circle--link"></i>
                </span>
              </td>
              <td className="u-align--right">
                {memoryWithUnit(general.free)}
                <span className="u-nudge-right--small">
                  <i className="p-circle--link-faded"></i>
                </span>
              </td>
            </tr>
            {totalHugepages > 0 && (
              <tr data-test="hugepages-data">
                <td>
                  Hugepage
                  {hugepages.pageSize && (
                    <>
                      <br />
                      <strong
                        className="p-text--x-small u-text--light"
                        data-test="page-size"
                      >
                        {`(Size: ${memoryWithUnit(hugepages.pageSize)})`}
                      </strong>
                    </>
                  )}
                </td>
                <td className="u-align--right">
                  {memoryWithUnit(hugepages.allocated)}
                  <span className="u-nudge-right--small">
                    <i className="p-circle--positive"></i>
                  </span>
                </td>
                <td className="u-align--right">
                  {memoryWithUnit(hugepages.free)}
                  <span className="u-nudge-right--small">
                    <i className="p-circle--positive-faded"></i>
                  </span>
                </td>
              </tr>
            )}
          </tbody>
        </table>
      </div>
    </div>
  );
};

export default RamResources;
