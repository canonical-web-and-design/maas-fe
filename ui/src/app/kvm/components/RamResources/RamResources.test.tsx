import { shallow } from "enzyme";

import RamResources from "./RamResources";

import { COLOURS } from "app/base/constants";

describe("RamResources", () => {
  it("renders", () => {
    const wrapper = shallow(
      <RamResources
        generalAllocated={1}
        generalFree={2}
        generalOther={3}
        hugepagesAllocated={4}
        hugepagesFree={5}
        hugepagesOther={6}
        pageSize={7}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("can be made to have a dynamic layout", () => {
    const wrapper = shallow(
      <RamResources dynamicLayout generalAllocated={1} generalFree={2} />
    );

    expect(
      wrapper
        .find(".ram-resources")
        .prop("className")
        ?.includes("ram-resources--dynamic-layout")
    ).toBe(true);
  });

  it("shows hugepages data if provided", () => {
    const wrapper = shallow(
      <RamResources
        generalAllocated={1}
        generalFree={2}
        hugepagesAllocated={3}
        hugepagesFree={4}
      />
    );

    expect(wrapper.find("[data-testid='hugepages-data']").exists()).toBe(true);
  });

  it("does not show hugepages data if not provided", () => {
    const wrapper = shallow(
      <RamResources generalAllocated={1} generalFree={2} />
    );

    expect(wrapper.find("[data-testid='hugepages-data']").exists()).toBe(false);
  });

  it("show hugepages page size if provided", () => {
    const wrapper = shallow(
      <RamResources
        generalAllocated={1}
        generalFree={2}
        hugepagesAllocated={3}
        hugepagesFree={4}
        pageSize={5}
      />
    );

    expect(wrapper.find("[data-testid='page-size']").exists()).toBe(true);
    expect(wrapper.find("[data-testid='page-size']").text()).toBe("(Size: 5B)");
  });

  it("does not show hugepages page size if not provided", () => {
    const wrapper = shallow(
      <RamResources
        generalAllocated={1}
        generalFree={2}
        hugepagesAllocated={3}
        hugepagesFree={4}
      />
    );

    expect(wrapper.find("[data-testid='page-size']").exists()).toBe(false);
  });

  it("can show whether RAM has been over-committed", () => {
    const wrapper = shallow(
      <RamResources
        generalAllocated={1}
        generalFree={-1}
        hugepagesAllocated={3}
        hugepagesFree={-1}
      />
    );

    expect(wrapper.find("DoughnutChart").prop("segments")).toStrictEqual([
      { color: COLOURS.CAUTION, value: 1 },
    ]);
  });

  it("shows RAM used by other projects if data provided", () => {
    const wrapper = shallow(
      <RamResources generalAllocated={1} generalFree={2} generalOther={3} />
    );

    expect(wrapper.find("[data-testid='others-col']").exists()).toBe(true);
  });
});
