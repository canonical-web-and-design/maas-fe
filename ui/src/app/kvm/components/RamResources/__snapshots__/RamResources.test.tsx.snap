// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RamResources renders 1`] = `
<div
  className="ram-resources"
>
  <div
    className="ram-resources__chart-container"
  >
    <h4
      className="p-heading--small"
    >
      RAM
    </h4>
    <DoughnutChart
      className="ram-resources__chart"
      label="21B"
      segmentHoverWidth={18}
      segmentWidth={15}
      segments={
        Array [
          Object {
            "color": "#0066CC",
            "tooltip": "Allocated: 1B general + 4B hugepages",
            "value": 5,
          },
          Object {
            "color": "#0E8420",
            "tooltip": "Others: 3B general + 6B hugepages",
            "value": 9,
          },
          Object {
            "color": "#D3E4ED",
            "tooltip": "Free: 2B general + 5B hugepages",
            "value": 7,
          },
        ]
      }
      size={96}
    />
  </div>
  <div
    className="ram-resources__table-container"
  >
    <table
      className="ram-resources__table"
    >
      <thead>
        <tr>
          <th />
          <th
            className="u-align--right u-text--light u-truncate"
          >
            Allocated
            <span
              className="u-nudge-right--small"
            >
              <i
                className="p-circle--link"
              />
            </span>
          </th>
          <th
            className="u-align--right u-text--light u-truncate"
            data-testid="others-col"
          >
            Others
            <span
              className="u-nudge-right--small"
            >
              <i
                className="p-circle--positive"
              />
            </span>
          </th>
          <th
            className="u-align--right u-text--light u-truncate"
          >
            Free
            <span
              className="u-nudge-right--small"
            >
              <i
                className="p-circle--link-faded"
              />
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            General
          </td>
          <td
            className="u-align--right"
          >
            <span
              className="u-nudge-left"
            >
              1B
            </span>
          </td>
          <td
            className="u-align--right"
          >
            <span
              className="u-nudge-left"
            >
              3B
            </span>
          </td>
          <td
            className="u-align--right"
          >
            <span
              className="u-nudge-left"
            >
              2B
            </span>
          </td>
        </tr>
        <tr
          data-testid="hugepages-data"
        >
          <td>
            Hugepage
            <br />
            <strong
              className="p-text--x-small u-text--light"
              data-testid="page-size"
            >
              (Size: 7B)
            </strong>
          </td>
          <td
            className="u-align--right"
          >
            <span
              className="u-nudge-left"
            >
              4B
            </span>
          </td>
          <td
            className="u-align--right"
          >
            <span
              className="u-nudge-left"
            >
              6B
            </span>
          </td>
          <td
            className="u-align--right"
          >
            <span
              className="u-nudge-left"
            >
              5B
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
